rfd/0164-scoped-rbac.md

```diff
@@ -0,0 +1,1122 @@
+that will evaluate grants assigned to the identity at the matching scope and access request ids.
+
+### The "New enemy" problem
+
+The Zanzibar paper describes two cases of "new enemy problem" (https://research.google/pubs/zanzibar-googles-consistent-global-authorization-system/):
+
+**Case 1 - Neglecting update order**
+
+* Step 1. Alice removes Bob from the ACL of a folder.
+* Step 2. Alice asks Charlie to add new docs to a folder.
+* Step 3. Bob should not see new documents, but can do so, if the ACL neglects the ordering. (Because in Zanzibar, individual documents may inherit ACL of the parent folder).
+
+**Case 2 - Misapplying old ACL to new content**
+
+* Step 1. Alice removes Bob from the ACL of the document.
+* Step 2. Alice asks Charlie to add new contents to the document.
+* Step 3. Bob should not be able to see new content, but may do so, if ACL evaluates Bob's permissions at Step 2 before reading Step 1.
+
+Let's review how those cases can apply to Teleport's Grants.
+
+Bob was granted role `access` in scope `/dev/lab`.
+
+**Case 1 - Neglecting update order** 
+
+* Step 1. Alice removes Bob from the access list granting `access` in scope of `/dev/lab`
+* Step 2. Alice adds new resource `luna` to the scope of `/dev/lab`
+* Step 3. Bob should not get access to `luna`, but can do so, if the Teleport Proxy evaluating access skips `Step 1` before evaluating `Step 2`
+
+Let's assume that every time a resource is added or removed to a resource group, it has it's timestamp updated.
+
+* Before any of the step, we can assume that grant hierarchy `G` - `/dev/lab` has some timestamp `Tg` and a resource hierarchy - `R` has some timestamp `Tr`.
+* At a `Step 1`, at the timestamp `T1` Alice removed Bob from access list granting role `access` in the scope `/dev/lab`. Teleport has updated the `_root` nodes of grant hierarchy `/dev` and `/dev/lab` with this timestamp in the backend.
+* At a `Step 2`, at the timestamp `T2` > `T1`, Alice added a new resource `luna` to the scope of the `/dev/lab`. Teleport has updated the `_root` node of the resource group hierarchy `/dev` and `/dev/lab` with this timestamp in the backend.
+* At a `Step 3`, Teleport proxy would have to evaluate access to host luna. The proxy would have to make sure that it's copy of the grant hierarchies `/dev/lab` is not stale.
+
+In our example `T2 > T1 > Tr` and `T2 > T1 > Tg`. There are two sub cases when `Tr > Tg` and `Tr < Tg`
+
+Here are possible combinations of versions that The Proxy can encounter:
+
+```
+1. G - Tg, R - Tr
+2. G - T1, R - Tr
+3. G - Tg, R - T2
+4. G - T1, R - T2
+```
+
+Let's review the cases in the reverse order:
+
+* In the cases 3 and 4, the Proxy sees that the resource group has been updated after the grant, because `T2 > T1 > Tg` and can proceed.
+* In the case 2, The proxy sees that `R` may be stale, because it's `Tr` less than `T1 and has to make a call to the backend making sure it fetched update of `R` least as fresh as `T1` with membership of `luna`.
```

Comment:
```suggestion
* In the case 2, The proxy sees that `R` may be stale, because it's `Tr` less than `T1` and has to make a call to the backend making sure it fetched update of `R` least as fresh as `T1` with membership of `luna`.
```